src/HydroShocktubeCMA/test_hydro_shocktube_cma.cpp

```diff
@@ -0,0 +1,295 @@
+
+		consVar(i, j, k, HydroSystem<ShocktubeProblem>::scalar0_index + 0) = 0.8;
+		consVar(i, j, k, HydroSystem<ShocktubeProblem>::scalar0_index + 1) = 0.3 * pow(sin(20 * 3.14 * 0), 2);
+		consVar(i, j, k, HydroSystem<ShocktubeProblem>::scalar0_index + 2) = 1 - 0.8 - 0.3 * pow(sin(20 * 3.14 * 0), 2);
+
+	} else if (i >= hi[0]) {
+		// x1 right-side boundary -- constant
+		for (int n = 0; n < numcomp; ++n) {
+			consVar(i, j, k, n) = 0;
+		}
+
+		consVar(i, j, k, RadSystem<ShocktubeProblem>::gasEnergy_index) = P_R / (gamma - 1.);
+		consVar(i, j, k, RadSystem<ShocktubeProblem>::gasInternalEnergy_index) = P_R / (gamma - 1.);
+		consVar(i, j, k, RadSystem<ShocktubeProblem>::gasDensity_index) = rho_R;
+		consVar(i, j, k, RadSystem<ShocktubeProblem>::x1GasMomentum_index) = 0.;
+		consVar(i, j, k, RadSystem<ShocktubeProblem>::x2GasMomentum_index) = 0.;
+		consVar(i, j, k, RadSystem<ShocktubeProblem>::x3GasMomentum_index) = 0.;
+
+		consVar(i, j, k, HydroSystem<ShocktubeProblem>::scalar0_index + 0) = 0.1;
+		consVar(i, j, k, HydroSystem<ShocktubeProblem>::scalar0_index + 1) = 0.3 * pow(sin(20 * 3.14 * 1), 2);
+		consVar(i, j, k, HydroSystem<ShocktubeProblem>::scalar0_index + 2) = 1 - 0.1 - 0.3 * pow(sin(20 * 3.14 * 1), 2);
+	}
+}
+
+template <> void RadhydroSimulation<ShocktubeProblem>::ErrorEst(int lev, amrex::TagBoxArray &tags, Real /*time*/, int /*ngrow*/)
+{
+	// tag cells for refinement
+
+	const Real eta_threshold = 0.1; // gradient refinement threshold
+	const Real rho_min = 0.01;	// minimum rho for refinement
+	auto const &dx = geom[lev].CellSizeArray();
+
+	for (amrex::MFIter mfi(state_new_cc_[lev]); mfi.isValid(); ++mfi) {
+		const amrex::Box &box = mfi.validbox();
+		const auto state = state_new_cc_[lev].const_array(mfi);
+		const auto tag = tags.array(mfi);
+
+		amrex::ParallelFor(box, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+			int const n = 0;
+			Real const rho = state(i, j, k, n);
+			Real const del_x = (state(i + 1, j, k, n) - state(i - 1, j, k, n)) / (2.0 * dx[0]);
+			Real const gradient_indicator = std::sqrt(del_x * del_x) / rho;
+
+			if (gradient_indicator > eta_threshold && rho >= rho_min) {
+				tag(i, j, k) = amrex::TagBox::SET;
+			}
+		});
+	}
+}
+
+template <> void RadhydroSimulation<ShocktubeProblem>::computeAfterTimestep()
+{
+	auto [position, values] = fextract(state_new_cc_[0], Geom(0), 0, 0.5);
+	const int nx = static_cast<int>(position.size());
+	const int nmscalars = Physics_Traits<ShocktubeProblem>::numMassScalars;
+
+	if (amrex::ParallelDescriptor::IOProcessor()) {
+
+		for (int nn = 0; nn < nx; ++nn) {
+			amrex::Real specieSum = 0.0;
+			for (int n = 0; n < nmscalars; ++n) {
+				specieSum += values.at(HydroSystem<ShocktubeProblem>::scalar0_index + n)[nn];
+				amrex::Print() << "nn = " << nn << " n = " << n << " val = " << values.at(HydroSystem<ShocktubeProblem>::scalar0_index + n)[nn]
+					       << " \n";
+			}
+			amrex::Print() << "delta = " << 1e0 - specieSum << "\n";
```

Comment:
I don't think the scalars should add to 1 in every cell. I think what you want to do is to compute the error in `specieSum` relative to the gas density. If you extract the gas density as `amrex::Real rho = values.at(HydroSystem<ShocktubeProblem>::density)[nn];` then you can compute the relative error as `amrex::Real err = (specieSum - rho)/rho;`

Reply:
Why? In the initial conditions, the values of mass scalars I give is their mass fractions, and they sum up to unity. Shouldn't at every timestep their sum be unity, as long as `scalar0_index +0/1/2` store the mass fractions?

Reply:
That is wrong. The mass scalars must be partial densities, i.e. their mass fractions multiplied by the gas density. The CMA algorithm and the Riemann solver both assume this.

Reply:
Woah. In FLASH, we always provided mass fractions, but then FLASH's MassScalars are separate from its chemical species, and the CMA conservation I implemented was for chemical species (Christoph extended it to MassScalars later but I never used them). Good we caught the issue now, because this means we need to change the values of mass scalars in the primordial chem test as well (currently, they are just mass fractions and not partial densities).

Reply:
It only matters if there's advection, so it won't change the answer of a one zone test. But I agree it should probably be changed in the other test for consistency.